taskgroup.go

```diff
@@ -10,24 +10,21 @@ import "sync"
 // tasks are collected and reported by the group.
 type Task func() error
 
-// A Group manages a collection of cooperating goroutines.  New tasks are added
-// to the group with the Go method.  Call the Wait method to wait for the tasks
-// to complete. A zero value is ready for use, but must not be copied after its
+// A Group manages a collection of cooperating goroutines.  Add new tasks to
+// the group with the Go method.  Call the Wait method to wait for the tasks to
+// complete. A zero value is ready for use, but must not be copied after its
 // first use.
 //
 // The group collects any errors returned by the tasks in the group. The first
 // non-nil error reported by any task (and not otherwise filtered) is returned
 // from the Wait method.
 type Group struct {
-	wg  sync.WaitGroup // counter for active goroutines
-	err error          // error returned from Wait
+	wg      sync.WaitGroup // counter for active goroutines
+	onError ErrorFunc      // called each time a task returns non-nil
 
-	setup sync.Once // set up and start the error collector
-	reset sync.Once // stop the error collector and set err
-
-	onError func(error) error // called each time a task returns non-nil
-	errc    chan<- error      // errors generated by goroutines
-	edone   chan struct{}     // signals error completion
+	μ     sync.Mutex // guards the fields below
+	setup sync.Once  // set up and start the error collector
```

Comment:
Hmm, I think this is safe but the docstrings here lie: access to setup is not guarded by the mutex in g.init, and so by extension err access isn't always guarded by the mutex either.

However... I believe it's true that, in the current implementation, all codepaths that might lead to grabbing mu and touching error must have gone through an earlier g.init, and so g.init's unguarded access always happens-before all accesses via mu+err.

This is still unsafe if someone violates the contract on Wait, because a concurrent call to Go would run g.init and touch g.setup while Wait is replacing g.setup. I wonder if you could get rid of setup and g.init at this point, and just reset err=nil while holding the mutex in Wait? I think that would end up with the same semantics throughout, and make some sharp edges go away.

Reply:
It is safe, but you're right that it needs more documentation. Specifically _calling_ the setup does not require holding the lock, but modifying the `setup` field itself does. I will expand the comment.

Reply:
You can't reset `g.err` in Wait without disturbing the property that if you have multiple things Wait'ing, they should get the same result (and also if you call wait again, etc.).

On the other hand, maybe that's not important enough to preserve. I think all the cases where I use this in anger, it's with a single Wait'er.

Reply:
(reworked as discussed offline)